### ISSUE TITLE
Group.get needs to allow parameterized local filenames to avoid overwriting

### ISSUE THREAD

**Issue by cooperlees** [fabric/fabric#1868]
Hi,

I'd like to be able to .get() on a ThreadingGroup and have the hostname/ip/somethingunique added to the `dst` filename so I can get the same file from multiple hosts.

Sorry if I missed it, but I can not see a clean way to move to:
```
ThreadingGroup(*hosts, **DEFAULT_FAB_KWARGS).get(self._src, self._dst)
```
from
```
        ...
        with fab.settings(fab.hide("running"), warn_only=True):
            fab.execute(self._get, hosts=hosts)

    @fab.parallel
    def _get(self):
        fab.get(self._src, self._dst + "." + fab.env.host)
```

Please merge or propose workaround if they exist! Thanks.

**Comment by bitprophet** [comment]
I'm hoping this is in the upgrade doc because yes, it's a clear missing feature from v1 :)

EDIT: it is not in that doc! I'll add a line item and I think this may be the "TODO" ticket for this particular aspect of the feature so I'll tweak the desc.

**Comment by bitprophet** [comment]
We'll probably want to honor how v1 did this (which I was just in the middle of reinventing myself before I edited this comment...sigh) which is tl;dr an interpolated string config value, handed a bunch of values for parameterization (remote basename and pathname, connection object params, etc).

Main diff would be that we should hand in just a `connection` key/val by itself instead of manually teasing out useful `Connection` obj attributes ourselves as individual keys (i.e. no `"{host}"`, but instead, `"{connection.host}"`).

**Comment by cooperlees** [comment]
Thanks. I checked the document multiple times for this feature and couldn't find it anywhere. Many thanks for acknowledging and updating the PR + docs.

I'm not fussed, long as there is a nice way with the new API to use it I'll be super happy. Also happy to try and help, but prob getting up to speed might take a bit of time.

**Comment by bitprophet** [comment]
Linking this to #1810 (Group.get/put) where it obviously becomes pretty necessary. Will see what shakes out. I'd prefer to solve it generically but may end up having to solve it for Group specifically first.

**Comment by bitprophet** [comment]
Thinking out loud. v1's behavior:

- interpolation is a mix of what in v2 would be Connection properties (host string, as a single mashed up thing) and remote file properties (dirname, basename and full path)
  - I do not recall offhand how much of the latter is trivially available in Paramiko and how much is being done on Fabric's end
- local path defaults to `<host-string>/<remote-path>`, so eg saying `get('/var/log/foo.log')` on a host  downloads to `remotehostname/var/log/foo.log` locally
  - This is true even for a single connection, i.e., it is more complex than one might assume for simple cases, in order to not change behavior for complex ones (or eg to overwrite the same file N times, which would always be irritating/surprising)

My gut says that for an initial port to v2, I want to do a MVP and not have it take a bunch of extra time to be 100%, if we can just add the remainder later on in compatible fashion. We also want to take the opportunity to revisit the above decisions.

- Currently, our `Transfer.get` defaults to the equivalent of "just the remote filename in local cwd", so we cannot reasonably change that without breaking current compat.
  - Though we could at least make `Group.get` (from #1810) do it by default, as that is new.
  - Long term (eg 3.x) it seems reasonable to make even `Connection/Transfer.get` do the "default to including remote hostname in local path" thing?
    - but I guess we'll see, if nobody clamours for it during the lifetime of 2.x...maybe it was an overthinking?
- Giving the interpolation the connection instance seems obvious, one can just do eg `"{host}/foo.log"` for hostname
  - Working in remote path bits is likely just an implementation detail one way or another
  - Arguably we want to be zen of python and make the user explicitly say `"{connection.host}/{file.basename}"` or whatnot, but I'm not sure that's worth the hassle for the average use case...? will see how it plays out in tests
- It's not clear to me how valuable the remote basename/pathname/fullpath bits are in reality, so if they require a bunch of fabric-level code vs already being trivially accessible from the Paramiko sftp objects, I'm very likely to make that the next iteration instead
  - or not at all - my gut says most use cases will be happy with per-connection directories or partial filenames, plus the remote basename
  - that said, since we clearly already have access to the base filename, having that part available seems a useful middle ground

**Comment by bitprophet** [comment]
That's a lot of words. What is the shakeout?

- Clear first step: interpolate some subset of connection attributes & remote file attributes, in Connection
- Clear second step: Group.get adds a default local path value containing Connection.host (relying on current 'if you gave a dir, the file goes inside it' behavior for the file part) 
- Unclear third step: auto-mkdir?
  - Tension between user friction (making your own directories is 💩 ) and "more code means more bugs" (autocreating directories leads to more possibilities for bugs/errors the user cannot work around) - but I suspect convenience wins here
  - Also tension between doing that magic in Connection vs in Group - I suspect having Connection do it is cleaner despite being very technically backwards incompatible.

**Comment by bitprophet** [comment]
Implemented the above. There's no useful pathy bits in Paramiko and we were already applying `posixpath.basename`, so I just extended that to give `dirname` and `basename` for the time being. Between those and the Connection bits, should be a pretty big upgrade for v2 / be about on par with v1.

I did also do auto-mkdir using `os.makedirs`. It's not ideal because we cannot easily track whether that call DID anything and therefore we cannot easily undo it if the transfer fails; so there will sometimes be extra directories lying around in those cases. Not worth fussing over for now.
